spring-cloud-azure-keyvault-config/src/main/java/com/microsoft/azure/spring/cloud/keyvault/config/auth/AuthenticationExecutor.java

```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE in the project root for
+ * license information.
+ */
+
+package com.microsoft.azure.spring.cloud.keyvault.config.auth;
+
+import com.microsoft.aad.adal4j.AuthenticationContext;
+import com.microsoft.aad.adal4j.AuthenticationResult;
+
+import java.util.concurrent.Future;
+
+public interface AuthenticationExecutor {

```

Comment:
1. How about use functional programming to represent actual acquire token method? Then you don't need such inheritance hierarchy.
2. Suggest to use `CompletableFuture` rather `Future`. The later only support blocking get.

Reply:
>1. How about use functional programming to represent actual acquire token method? Then you don't need such inheritance hierarchy.
Sure. Let me try it.
>2. Suggest to use CompletableFuture rather Future. The later only support blocking get.
It comes from ADAL.  We are just wrapping it for different credentials.

Reply:
After offline sync, leave the implementation as it is for better test-ability.